packages/animations/lib/src/composite_animation_widget.dart

```diff
@@ -0,0 +1,270 @@
+/// composition and only changing animation values of both transitions instead
+/// of changing the shape of widget subtree levels in [AnimatedBuilder]
+/// based on [AnimationStatus] that causes the state object in that particular
+/// subtree level to get recreated.
+///
+/// To ensure the state of the child widget will survive, don't define
+/// conditional logic to switch subtrees in both transition builders required
+/// by this widget.
+class CompositeAnimationWidget extends StatefulWidget {
+  /// Creates a widget that animate it's child with different transitions based
+  /// on [AnimationStatus] of [animation].
+  ///
+  /// If the ([animation]) value goes forward, only [forwardTransitionBuilder]
+  /// animates from 0.0 to 1.0 and If goes reverse, only [reverseTransitionBuilder]
+  /// animates from 1.0 to 0.0.
+  ///
+  /// [animation] is typically an [AnimationController] that drives the
+  /// transitions. [animation], [forwardTransitionBuilder] and
+  /// [reverseTransitionBuilder] cannot be null.
+  const CompositeAnimationWidget({
+    Key key,
+    @required this.animation,
+    @required this.forwardTransitionBuilder,
+    @required this.reverseTransitionBuilder,
+    this.child,
+    this.visibleAtStart = false,
+  })  : assert(animation != null),
+        assert(forwardTransitionBuilder != null),
+        assert(reverseTransitionBuilder != null),
+        super(key: key);
+
+  /// The animation that drives the [child]'s entrance and exit.
+  final Animation<double> animation;
+
+  /// The widget below this widget in the tree.
+  ///
+  /// This widget will transition in and out as driven by [animation].
+  final Widget child;
+
+  /// A builder that wraps [child] with an animation that will only play when
+  /// [animation] goes forward.
+  ///
+  /// To ensure the state of the child widget will survive, don't define this
+  /// builder with conditional logic to switch subtrees.
+  final ComponentTransitionBuilder forwardTransitionBuilder;
+
+  /// A builder that wraps [child] with an animation that will only play when
+  /// [animation] goes reverse.
+  ///
+  /// To ensure the state of the child widget will survive, don't define this
+  /// builder with conditional logic to switch subtrees.
+  final ComponentTransitionBuilder reverseTransitionBuilder;
+
+  /// A flag to determine if you defined your transition builders
+  /// ([forwardTransitionBuilder] and
+  /// [reverseTransitionBuilder]) to have a
+  /// "visible-point" of animation at 0.0 ([AnimationStatus.dismissed]).
+  ///
+  /// This is usually set to true if you want to animate a widget to
+  /// disappear but with different exit and reenter animation without enter
+  /// animation (visible at start), which means you defined exit animation at
+  /// [forwardTransitionBuilder] and reenter animation
+  /// at [reverseTransitionBuilder].
+  ///
+  /// Some transition implementations like [FadeThroughTransition] and
```

Comment:
It's not entirely clear to me why we need this setting for the FadeThroughTransition and SharedAxisTransition. The FadeThroughTransition and SharedAxisTransition should just define their forwardTransitionBuilder/reverseTransitionBuilder properly.

Reply:
To better understanding, lets take a look at this table
![compo](https://user-images.githubusercontent.com/26335640/81863310-94f96d80-959d-11ea-9c06-9c94f1c0be89.png)
The use of `visibleAtStart` explicitly telling the widget that we want to define an "exit"/"disappear" animation, thus the secondary animation part of some page transition widgets. This is also the reason why I renamed those builders to avoid confusion like defining an "exit" animation in `enterBuilder` and "reenter" animation in `exitBuilder`.

On the technical side, we want the widget to know where those "values" stays while the another builder is animating. If we don't set the flag correctly, it will end up the builder to animate but we cant see the actual animation since the all the animation widget defined are built by composition being the `forwardBuilder` is the parent and the child is the `reverseBuilder`. For example, if the `reverseBuilder`'s `FadeTransition` opacity stays at '0.0' and the `forwardBuilder`'s `ScaleTransition` is animating, we can't see the animation. 

```
return widget.forwardTransitionBuilder(
          widget.reverseTransitionBuilder(
            child,
            _reverseAnimation,
          ),
          _forwardAnimation,
        );
```

`visibleAtStart` sets the x values to `0.0` if true, else '1.0'.

With your suggestion about `reverseBuilder` must run forward, I can simply add a flip function to `_reverseAnimation` .